src/rpc/blockchain.cpp

```diff
@@ -579,34 +580,38 @@ static RPCHelpMan getblockheader()
     };
 }
 
-static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex* pblockindex)
 {
-    AssertLockHeld(::cs_main);
     CBlock block;
-    if (blockman.IsBlockPruned(pblockindex)) {
-        throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+    {
+        LOCK(cs_main);
+        if (blockman.IsBlockPruned(pblockindex)) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+        }
     }
 
     if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus())) {
         // Block not found on disk. This could be because we have the block
         // header in our index but not yet have the block or did not accept the
-        // block.
+        // block. Or if the block was pruned right after we released the lock above.
```

Comment:
Do we have a test for such a case?

Reply:
I don't think it it possible to write a test for this deletion race, unless you want to modify the source code as well?

It might be easier to write a functional test for lack of deletion (on Windows) in https://github.com/bitcoin/bitcoin/pull/26533 by holding a open file handle to the block file

Reply:
The way I could see a test for this case would be to have a long sleep above the `if` statement, then call `PruneBlockFilesManual` in the test and it should fail here. Is there another way to introduce an arbitrary delay or block the thread in tests?